src/asfvideo.cpp

```diff
@@ -0,0 +1,770 @@
+  int i;
+  for (i = 0; i < 4; ++i) {
+    GUID[(3 - i) * 2] = returnHEX(buf[i] / 0x10);
+    GUID[(3 - i) * 2 + 1] = returnHEX(buf[i] % 0x10);
+  }
+  for (i = 4; i < 6; ++i) {
+    GUID[(9 - i) * 2 + 1] = returnHEX(buf[i] / 0x10);
+    GUID[(9 - i) * 2 + 2] = returnHEX(buf[i] % 0x10);
+  }
+  for (i = 6; i < 8; ++i) {
+    GUID[(14 - i) * 2] = returnHEX(buf[i] / 0x10);
+    GUID[(14 - i) * 2 + 1] = returnHEX(buf[i] % 0x10);
+  }
+  for (i = 8; i < 10; ++i) {
+    GUID[i * 2 + 3] = returnHEX(buf[i] / 0x10);
+    GUID[i * 2 + 4] = returnHEX(buf[i] % 0x10);
+  }
+  for (i = 10; i < 16; ++i) {
+    GUID[i * 2 + 4] = returnHEX(buf[i] / 0x10);
+    GUID[i * 2 + 5] = returnHEX(buf[i] % 0x10);
+  }
+  GUID[36] = '\0';
+  GUID[8] = GUID[13] = GUID[18] = GUID[23] = '-';
+}
+
+/*!
+  @brief Function used to check if data stored in buf is equivalent to
+      ASF Header TagVocabulary's GUID.
+  @param buf Exiv2 byte buffer
+  @return Returns true if the buffer data is equivalent to Header GUID.
+ */
+bool isASFType(byte buf[]) {
+  if (buf[0] == 0x30 && buf[1] == 0x26 && buf[2] == 0xb2 && buf[3] == 0x75 && buf[4] == 0x8e && buf[5] == 0x66 &&
+      buf[6] == 0xcf && buf[7] == 0x11 && buf[8] == 0xa6 && buf[9] == 0xd9 && buf[10] == 0x00 && buf[11] == 0xaa &&
+      buf[12] == 0x00 && buf[13] == 0x62 && buf[14] == 0xce && buf[15] == 0x6c)
+    return true;
+
+  return false;
+}
+
+//! Function used to convert buffer data into 64-bit Integer, information stored in littleEndian format
+uint64_t getUint64_t(Exiv2::DataBuf& buf) {
+  uint64_t temp = 0;
+
+  for (int i = 0; i < 8; ++i) {
+    temp = temp + static_cast<uint64_t>(buf.data()[i] * (pow(static_cast<float>(256), i)));
+  }
+  return temp;
+}
+
+}  // namespace Internal
+}  // namespace Exiv2
+
+namespace Exiv2 {
+
+using namespace Exiv2::Internal;
+
+AsfVideo::AsfVideo(BasicIo::UniquePtr io) : Image(ImageType::asf, mdNone, std::move(io)) {
+}  // AsfVideo::AsfVideo
+
+std::string AsfVideo::mimeType() const {
+  return "video/asf";
+}
+
+void AsfVideo::writeMetadata() {
+}
+
+void AsfVideo::readMetadata() {
+  if (io_->open() != 0)
+    throw Error(ErrorCode::kerDataSourceOpenFailed, io_->path(), strError());
+
+  // Ensure that this is the correct image type
+  if (!isAsfType(*io_, false)) {
+    if (io_->error() || io_->eof())
+      throw Error(ErrorCode::kerFailedToReadImageData);
+    throw Error(ErrorCode::kerNotAnImage, "ASF");
+  }
+
+  IoCloser closer(*io_);
+  clearMetadata();
+  continueTraversing_ = true;
+  io_->seek(0, BasicIo::beg);
+  height_ = width_ = 1;
+
+  xmpData()["Xmp.video.FileSize"] = (double)io_->size() / (double)1048576;
+  xmpData()["Xmp.video.FileName"] = io_->path();
+  xmpData()["Xmp.video.MimeType"] = mimeType();
+
+  while (continueTraversing_)
+    decodeBlock();
+
+  aspectRatio();
+}  // AsfVideo::readMetadata
+
+void AsfVideo::decodeBlock() {
+  const long bufMinSize = 9;
+  DataBuf buf(bufMinSize);
+  uint64_t size = 0;
+  buf.data()[8] = '\0';
```

Comment:
[suggestion] Do not need to initialize end of buffer with `\0`. The buffer is already filled with null characters.

Reply:
Done here ad72e1e6b0c2a10383a3878c1e603aa78fe3275a